### PR TITLE
feat(extension): support for batch translation for LLM

### LINKED ISSUE #385: [FEATURE] 可以支持合并翻译吗
### Before you begin

- [x] I have searched existing issues to avoid duplicates
- [ ] I am willing to contribute a PR for this feature

### Is your feature request related to a problem? Please describe.

之前用的是电子书翻译，都是自带合并翻译的，就是将很多个段落统一到一起翻译。这样的好处不仅是上下文更长更准确，还在于非常节省调用次数。
现在的翻译是一个段落，几个字符就一次调用，一个网页就可以把很多API的当天调用次数额度用尽了。

### Describe the solution you'd like

合并

### Which area(s) would this feature affect? (Select all that apply)

Website, Browser Extension

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Yukiniro**
## Type of Changes

- [x] ✨ New feature (feat)
- [ ] 🐛 Bug fix (fix)
- [ ] 📝 Documentation change (docs)
- [ ] 💄 UI/style change (style)
- [ ] ♻️ Code refactoring (refactor)
- [ ] ⚡ Performance improvement (perf)
- [x] ✅ Test related (test)
- [ ] 🔧 Build or dependencies update (build)
- [ ] 🔄 CI/CD related (ci)
- [x] 🌐 Internationalization (i18n)
- [x] 🧠 AI model related (ai)
- [ ] 🔄 Revert a previous commit (revert)
- [ ] 📦 Other changes that do not modify src or test files (chore)

## Related Issue

Closes #385 

## How Has This Been Tested?

- [x] Added unit tests
- [x] Verified through manual testing

## Screenshots

image

## Checklist

- [x] I have tested these changes locally
- [ ] I have updated the documentation accordingly if necessary
- [ ] My code follows the code style of this project
- [ ] My changes do not break existing functionality
- [x] If my code was generated by AI, I have proofread and improved it as necessary.

## Additional Information

    

---

## Summary by cubic
Adds LLM batch translation that groups multiple paragraphs into a single request to reduce API calls and speed up translations. Fulfills #385 with configurable batch limits in settings.

- **New Features**
  - Added BatchQueue to group LLM translation tasks by config and flush by size, characters, or delay.
  - LLM translations use a batch prompt with the "%%" separator to return aligned per-paragraph results.
  - New “Batch Translation” settings for Maximum Characters and Maximum Paragraph Count.
  - Background queue routes LLM requests through BatchQueue; added message handler to update batch config.
  - Comprehensive unit tests for batching logic and edge cases; updated i18n strings across locales.

- **Migration**
  - Config schema v25 with auto-migration adding requestBatchConfig.
  - Defaults: 1000 characters, 4 paragraphs.

### CODE REVIEW

**Comment by Yukiniro** [comment]
@codex review

File: apps/extension/src/entrypoints/background/request-queue.ts [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 import { executeTranslate } from '@/utils/host/translate/translate-text'
 import { onMessage } from '@/utils/message'
+import { BatchQueue } from '@/utils/request/batch-queue'
 import { RequestQueue } from '@/utils/request/request-queue'
 import { ensureInitializedConfig } from './config'
 
 export async function setUpRequestQueue() {
   const config = await ensureInitializedConfig()
-  const { translate: { requestQueueConfig: { rate, capacity } } } = config ?? DEFAULT_CONFIG
+  const { translate: { requestQueueConfig: { rate, capacity }, requestBatchConfig: { batchCharacters, batchSize } } } = config ?? DEFAULT_CONFIG
```
Comment (mengxi-ream) [inline apps/extension/src/entrypoints/background/request-queue.ts]: rename `batchCharacters` to `maxCharactersPerBatch`, `batchSize` to `maxItemsPerBatch` to make the meaning more clear

File: apps/extension/src/entrypoints/background/request-queue.ts [hunk 2]
```diff
@@ -16,6 +18,12 @@ export async function setUpRequestQueue() {
     maxRetries: 2,
     baseRetryDelayMs: 1_000,
   })
+  const batchQueue = new BatchQueue({
```
Comment (mengxi-ream) [inline apps/extension/src/entrypoints/background/request-queue.ts]: if we use the name batchQueue here, the we should use `batchQueueConfig` above to make them consistent

File: apps/extension/src/utils/config/__tests__/example/v025.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
+import type { TestSeriesObject } from './types'
+import type { Config } from '@/types/config/config'
+
+export const testSeries: TestSeriesObject = {
+  'complex-config-from-v020': {
+    description: 'Add beta experience configuration',
```
Comment (mengxi-ream) [inline apps/extension/src/utils/config/__tests__/example/v025.ts]: change description here

File: apps/extension/src/utils/config/__tests__/example/v025.ts [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        width: 700,
+      },
+      selectionToolbar: { enabled: false },
+      betaExperience: {
+        enabled: false,
+      },
+    } satisfies Config,
+  },
+  'config-from-v022': {
+    description: 'Add beta experience configuration with disabled state',
```
Comment (mengxi-ream) [inline apps/extension/src/utils/config/__tests__/example/v025.ts]: change description here to reflect what we did in the newest v024-to-v025

File: apps/extension/src/utils/prompts/translate.ts [hunk 5]
```diff
@@ -13,3 +13,21 @@ export async function getTranslatePrompt(targetLang: string, input: string) {
     .replaceAll(getTokenCellText(TARGET_LANG), targetLang)
     .replaceAll(getTokenCellText(INPUT), input)
 }
+
+export async function getBatchTranslatePrompt(targetLang: string, input: string) {
```
Comment (mengxi-ream) [inline apps/extension/src/utils/prompts/translate.ts]: can merge it into getTranslatePrompt into one single function to reduce duplicate code so that in the future we can easily add more cells like `targetLang`, for example, in the future we will add `websiteTitle`, `websiteDescription`, and only need to change the code in one function if we merge `getTranslatePrompt` and `getBatchTranslatePrompt`

File: apps/extension/src/utils/message.ts [hunk 6]
```diff
@@ -24,6 +25,7 @@ interface ProtocolMap {
   // request
   enqueueTranslateRequest: (data: { text: string, langConfig: Config['language'], providerConfig: ProviderConfig, scheduleAt: number, hash: string }) => Promise<string>
   setTranslateRequestQueueConfig: (data: { rate?: number, capacity?: number }) => void
+  setTranslateRequestBatchConfig: (data: Partial<RequestBatchConfig>) => void
```
Comment (mengxi-ream) [inline apps/extension/src/utils/message.ts]: `Partial` is beautiful! could you refactor type of `setTranslateRequestQueueConfig` to `Partial<RequestQueueConfig>` as well?

btw, here we may need to change the name `setTranslateRequestBatchConfig` to `setTranslateBatchQueueConfig` and `RequestBatchConfig` to `BatchQueueConfig` to make it consistent with `BatchQueue` class name

File: apps/extension/src/utils/request/batch-queue.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+export interface BatchOptions {
+  batchCharacters: number
+  batchSize: number
+  batchDelay: number
+}
+
+export class BatchQueue {
+  private requestQueue: RequestQueue | null = null
+  private pendingBatchMap = new Map<string, PendingBatch>()
+  private nextScheduleTimer: NodeJS.Timeout | null = null
```
Comment (mengxi-ream) [inline apps/extension/src/utils/request/batch-queue.ts:34]: should we use `ReturnType<typeof setTimeout>` ot `NodeJS.Timeout`, I saw you change the type to `ReturnType<typeof setTimeout>` in `node-translate.ts`

File: apps/extension/src/utils/request/batch-queue.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  private batchCharacters: number
+  private batchSize: number
+  private batchDelay: number
+
+  constructor(config: BatchOptions) {
+    this.batchCharacters = config.batchCharacters
```
Comment (mengxi-ream) [inline apps/extension/src/utils/request/batch-queue.ts]: rename this to `maxCharactersPerBatch` and `maxItemsPerBatch`

File: apps/extension/src/utils/request/batch-queue.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    this.batchSize = config.batchSize
+    this.batchDelay = config.batchDelay
+  }
+
+  enqueue(text: string, langConfig: Config['language'], providerConfig: ProviderConfig, scheduleAt: number, hash: string): Promise<string> {
```
Comment (mengxi-ream) [inline apps/extension/src/utils/request/batch-queue.ts]: this design coupling translate with batch queue. In the future, we want to batch the reading requests using this queue as well. so need to change the design a little bit.

Let's discuss this later, I have some proposals for this.

File: apps/extension/src/utils/constants/translate.ts [hunk 10]
```diff
@@ -1,9 +1,20 @@
 export const MIN_TRANSLATE_RATE = 1
 export const MIN_TRANSLATE_CAPACITY = 1
+export const MIN_BATCH_CHARACTERS = 1
+export const MIN_BATCH_SIZE = 1
 
 export const DEFAULT_REQUEST_RATE = 8
 export const DEFAULT_REQUEST_CAPACITY = 60
 
+export const DEFAULT_BATCH_CHARACTERS = 1000
+export const DEFAULT_BATCH_MIN_BATCH_SIZE = 4
```
Comment (mengxi-ream) [inline apps/extension/src/utils/constants/translate.ts]: use `DEFALT_MAX_CHARACTER_PER_BATCH` and `DEFAULT_MAX_ITEMS_PER_BATCH`

File: apps/extension/src/utils/constants/translate.ts [hunk 11]
```diff
@@ -1,9 +1,20 @@
 export const MIN_TRANSLATE_RATE = 1
 export const MIN_TRANSLATE_CAPACITY = 1
+export const MIN_BATCH_CHARACTERS = 1
+export const MIN_BATCH_SIZE = 1
```
Comment (mengxi-ream) [inline apps/extension/src/utils/constants/translate.ts]: use `MIN_BATCH_ITEMS`

File: apps/extension/src/utils/request/types.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
+export interface BatchQueueOptions {
```
Comment (mengxi-ream) [inline apps/extension/src/utils/request/types.ts]: may need to remove this file? cuz we didn't use these defintions

File: apps/extension/src/entrypoints/background/request-queue.ts [hunk 13]
```diff
@@ -17,6 +30,31 @@ export async function setUpRequestQueue() {
...
+    getBatchKey: (data) => {
+      return Sha256Hex(`${data.langConfig.sourceCode}-${data.langConfig.targetCode}-${data.providerConfig.id}`)
+    },
+    getCharacters: (data) => {
+      return data.text.length
+    },
+    executeBatch: async (dataList) => {
+      const { langConfig, providerConfig } = dataList[0]
+      const texts = dataList.map(d => d.text)
+      const batchText = texts.join(BATCH_SEPARATOR)
```
Comment (mengxi-ream) [inline apps/extension/src/entrypoints/background/request-queue.ts]: Google Chrome 2025-09-30 14 28 30
didn't seperate the translation into new lines, it like 

```
A%%B%%C
```

But we expect this:

```
A
%%
B
%%
C
```

**Review by mengxi-ream — CHANGES_REQUESTED** [review]

File: apps/extension/src/locales/en.yml [hunk 14]
```diff
@@ -161,6 +161,15 @@ options:
       rate:
         title: Average Requests Per Second
         description: Average requests per second, but allows configuring a maximum burst count to temporarily send excess requests
+    batchQueueConfig:
+      title: Batch Translation
+      description: Configure character limits and maximum paragraph count for batch translation (only effective for LLM translation)
+      maxCharactersPerBatch:
+        title: Maximum Characters
+        description: Maximum number of characters allowed per batch request to control the content volume of each translation (only effective for LLM translation)
```
Comment (mengxi-ream) [inline apps/extension/src/locales/en.yml]: remove `(only effective for LLM translation)` here since you've mentioned this above

File: apps/extension/src/locales/en.yml [hunk 15]
```diff
@@ -161,6 +161,15 @@ options:
...
+      maxItemsPerBatch:
+        title: Maximum Paragraph Count
+        description: Maximum number of paragraphs for each batch translation to control the content volume of each translation (only effective for LLM translation)
```
Comment (mengxi-ream) [inline apps/extension/src/locales/en.yml]: remove `(only effective for LLM translation)` here since you've mentioned this above

**Review by mengxi-ream — APPROVED** [review]

**Review by mengxi-ream — CHANGES_REQUESTED** [review]
I will add the retry logic
